### PR TITLE
PSBT base64 (de)serialization with Display & FromStr

### PR DESCRIPTION
**Pull request by dr-orlovsky**
Was quite surprised that the lib can't serialize and read Base64 PSBTs as required by BIP-174 (while having base64 as a dependency). Since this is a way of putting PSBT into a string as defined by the standard I used it as a way for implementing `FromStr/Display`.

~~Seems like an API break since we need to add one `psbt::Error` variant (feature-gated, though).~~

~~Provided one other commit https://github.com/rust-bitcoin/rust-bitcoin/pull/557/commits/a6399e7350a1b2ec6f91fc64001239402f4164ae that makes the PR non-API-breaking, suitable for `0.26.1` (really like to have base64 encoding ASAP). After the minor release that commit can be reverted.~~

### CODE REVIEW

File: src/util/psbt/mod.rs
Comment (sgeisler) [inline src/util/psbt/mod.rs:103]: This allocation is avoidable by using [`base64::display::Base64Display`](https://docs.rs/base64-compat/1.0.0/base64/display/index.html). I'm not sure if it's worth the inconsistency with decoding, but wanted to point out that option.

File: src/util/psbt/mod.rs
Comment (sgeisler) [inline src/util/psbt/mod.rs:112]: Swallowing the underlying error is suboptimal since it's unclear if it came from the base64 or from parsing the binary data structure (unless it is only used here which I doubt given the description and pre-existance).

**Review by sgeisler — COMMENTED** [review]
I left some nits, more descriptive errors are worth an API break imo.

EDIT: I just saw that a6399e7  is its own commit :smile: is it so important to you to keep the API stable to get into a point release? Seems hacky and we should definitely open an issue on merge or revert right away after the point release.

File: src/util/psbt/mod.rs
Comment (sgeisler) [reply src/util/psbt/mod.rs:112]: addressed in the commit comment of a6399e7

**Comment by dr-orlovsky** [comment]
@sgeisler thank you for the review! The reason for minor version is the following: 0.27 will take months to be published and will do a large-scale refactoring of the repo; while this thing is needed to be published within days/week, if possible.

**Comment by sgeisler** [comment]
As long as you take ownership of reverting the commit I'm ok with it, I'd just be unhappy with it staying that way and we don't exactly have a track record of being super organized as an org :smile:

**Comment by dr-orlovsky** [comment]
Improved display with your suggestion - and yes, I will create revert PR right after this one will be merged

**Review by sgeisler — APPROVED** [review]
ACK b407721

File: src/util/psbt/mod.rs
Comment (sanket1729) [inline src/util/psbt/mod.rs:475]: I used to `is_err` type checks before, but am shifting more towards asserting the error string equality. 
I realize that most of the repo probably has `is_err` checks everywhere, but we can at least make sure that incoming changes assert the correct variant.

**Review by sanket1729 — DISMISSED** [review]
tACK b4077213ab311bed2a4faaaae3600d9949b867a6. 

Left a minor nit, it will require some changes again in the revertme commit which are anyways going to be reverted. But I think that this is the right thing to do. I am not bent on it, as that is something related to repo in general and not something about this PR. 

We should also open an issue with milestone 0.27 so that we don't forget to revert the last commit.

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:475]: I am using `#[should_fail(expect="...")]` in my projects, but would not recommend it since tests break each time you proof/change description of the error

**Comment by dr-orlovsky** [comment]
@sanket1729 done with https://github.com/rust-bitcoin/rust-bitcoin/issues/564. I propose to merge this for now and discuss the best way of handling errors in that issue

File: src/util/psbt/mod.rs
Comment (sanket1729) [reply src/util/psbt/mod.rs:475]: I think that tests should fail if we change the description

**Comment by sanket1729** [comment]
@dr-orlovsky, I prefer not to merge unless I get a sign of approval from @apoelstra's powerful machine which runs more powerful checks than the CI, checks that every commit builds on the respective rustc versions that we care about. 

Maybe this is turning up too much work on @apoelstra and we need to find a better way. I am awaiting @apoelstra's comment before I merge this, anyways we need him for publishing the minor release.

**Comment by sgeisler** [comment]
@sanket1729 Are there any practical limits for on GH actions? I think we could test _a lot_ more if we wanted to, possibly gated by an initial smaller stage to not burn CPU needlessly.

**Comment by sanket1729** [comment]
@sgeisler, I think it also checks every single commit builds/tests on different rust versions which is I only do for the stable rust versions.

But we should see if we get per commit build/test checks in the CI script. Strangely enough, I can't easily find any project with CI that checks on every commit.

**Comment by sgeisler** [comment]
I tested b4077213ab311bed2a4faaaae3600d9949b867a6 quite extensively now (full feature matrix stable+nightly and multiple fuzzing rounds), I think we can just merge except if we want @apoelstra's ok regarding the release-and-revert strategy.

File: src/util/psbt/mod.rs
Comment (darosior) [inline src/util/psbt/mod.rs:464]: Hmm so you are not actually checking `psbt2`. `psbt` is already asserted above?

File: src/util/psbt/mod.rs
Comment (darosior) [inline src/util/psbt/mod.rs:475]: Same here

File: src/util/psbt/mod.rs
Comment (darosior) [inline src/util/psbt/mod.rs:486]: Same in all the tests actually

**Review by darosior — CHANGES_REQUESTED** [review]
The invalid vectors tests are wrong (they don't assert the actually deserialized from str version..)

File: src/util/psbt/mod.rs
Comment (stevenroose) [inline src/util/psbt/mod.rs:103]: I'd have putten the uses inside the base64 module, especially the consensus::encode one that's reused in both impls, but totally a minor nit.

**Review by stevenroose — CHANGES_REQUESTED** [review]
@darosior found some mistakes in the unit tests, so they should be fixed.

otherwise utACK

File: src/util/psbt/error.rs
Comment (apoelstra) [inline src/util/psbt/error.rs]: You cannot have a feature-gated error variant, in a major release or otherwise. This will always cause compilation breaks for downstream users that they cannot really control.

Comment (sgeisler) [reply src/util/psbt/error.rs]: Interesting, never thought about that, but probably because cargo uses the union of all features? So a new dependency that activates it can wreak havoc without exposing the relevant types, so it's legal for them to change the feature set without major version bumps.

Comment (dr-orlovsky) [reply src/util/psbt/error.rs]: @apoelstra yes, and this is exactly the reason why this change is not a part of this PR! As we discussing above, this PR includes last commit https://github.com/rust-bitcoin/rust-bitcoin/pull/557/commits/b4077213ab311bed2a4faaaae3600d9949b867a6 with which this is not an API break. This commit has to be reversed after the minor version release such that with the next major version we will get this error type improvement (but not before that).

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:103]: Will do once will understand with Andrew the strategy of merging such PR (with revert option for major API change)

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:486]: Thanks for catching this up. Will fix

File: src/util/psbt/error.rs
Comment (sgeisler) [reply src/util/psbt/error.rs]: Well, it appears we can't even do that. Let's imagine crate `A` depends on `rust-bitcoin` without features and crate `B` which also depends on rust-bitcoin without features, but doesn't expose it externally. Now `B` decides to turn on `base64` and does so in a minor release since it doesn't break the API (rust-bitcoin is a purely internal dependency). Now what cargo does is, when compiling one `rust-bitcoin` for both `A` and `B`, it calculates the union of all activated features in the dependency tree and uses that, breaking any previously exhaustive matches in `A`.

Comment (dr-orlovsky) [reply src/util/psbt/error.rs]: This will be the case when feature gate get introduced. In the current PR it is not: it will be there only after the revert of the last commit

Comment (sgeisler) [reply src/util/psbt/error.rs]: Well, afaik @apoelstra likes to review PRs commit-by-commit. Having this in there now means it won't be a clean revert but rather another change too. So there's really no point in having this feature-gating in the history and risking leaving it in when reverting.

**Comment by dr-orlovsky** [comment]
After a lot of confusion with the latest revert commit preventing API break, I have restructured the PR in this way:
- The current PR is **not API breaking** and reuses existing PSBT Error type, targeting 0.26.1 release
- It is followed by https://github.com/rust-bitcoin/rust-bitcoin/pull/574 which adds a custom error type specific for PSBT FromStr parsing, which is API breaking

File: src/util/psbt/error.rs
Comment (dr-orlovsky) [reply src/util/psbt/error.rs]: https://github.com/rust-bitcoin/rust-bitcoin/pull/557#issuecomment-782033192

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs]: Done as a separate commit, also improving error testing for previously existing tests: https://github.com/rust-bitcoin/rust-bitcoin/pull/557/commits/1209436a11dcedc4da0f9884c9ae10bb16e54370

**Review by sgeisler — DISMISSED** [review]
tACK fc006bb96fb5ad40fb70222b82cb141429a977b8

**Comment by dr-orlovsky** [comment]
Improved after discussion in https://github.com/rust-bitcoin/rust-bitcoin/pull/574#discussion_r579687827

**Comment by sgeisler** [comment]
Needs a fix for 1.29.0, but I like the error solution.

File: src/util/psbt/error.rs
Comment (sgeisler) [inline src/util/psbt/error.rs:103]: Is this still the case after introducing `PsbtParseError`?

File: src/util/psbt/mod.rs
Comment (sgeisler) [inline src/util/psbt/mod.rs:516]: Maybe I missed it before or you introduced it at a later point, but isn't this a horribly roundabout way of testing which error happened? Can we get rid of it after some of the error improvements got merged now (I assume you used this because our errors didn't have `Eq`)? Ideally we'd compare the returned result directly.

File: src/util/psbt/mod.rs
Comment (sgeisler) [inline src/util/psbt/mod.rs:104]: If you rebase on `master` this error can also derive `Clone, PartialEq, Eq`.

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:516]: I was not happy about this either, but there are two votes for introducing this thing (which was absent in my original version): https://github.com/rust-bitcoin/rust-bitcoin/pull/557#discussion_r567311296

File: src/util/psbt/error.rs
Comment (dr-orlovsky) [reply src/util/psbt/error.rs:103]: Yes, since it is about BIP-174 consensus encoding (PSBT binary encoding) and has nothing to do with Base64 or `FromStr` implementation

File: src/util/psbt/mod.rs
Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:104]: No, since I am limited by `consensus::encode::Error`: https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/consensus/encode.rs#L48

File: src/util/psbt/mod.rs
Comment (sgeisler) [reply src/util/psbt/mod.rs]: > I think that tests should fail if we change the description

I don't think this should be an explicit goal here. Since we can test for error-equality now that most of the derive-PRs are merged we should do that instead. If you feel strongly about error messages being tested that is probably better done in a separate test.

File: src/util/psbt/mod.rs
Comment (sgeisler) [reply src/util/psbt/mod.rs:516]: I disagree now that we can compare errors (before that the solution was certainly better than  `is_error()`).

File: src/util/psbt/mod.rs
Comment (sgeisler) [reply src/util/psbt/mod.rs:104]: Sorry, that's annoyning, I looked at the wrong error. So I guess this PR is as good as it gets without some breaking changes to make `consensus::encode::Error` implement `Eq`.

Comment (dr-orlovsky) [reply src/util/psbt/mod.rs:104]: All possible PRs to implement derives on Error types are already there: #559, #560, #558. Most of that is blocked by upstream. Overall tracking on Error derives across full bitcoin rust repos: https://github.com/orgs/rust-bitcoin/projects/2

File: src/util/psbt/mod.rs
Comment (sgeisler) [reply src/util/psbt/mod.rs:516]: irrelevant for now, because the relevant error can't actually impl `Eq` yet

**Review by sgeisler — APPROVED** [review]
tACK 7b7b73f6d2653d42be5bd63aa01a07680a7539af

<details>
  <summary>Test log:</summary>
  
`pr-557-head-7b7b73f6d2653d42be5bd63aa01a07680a7539af.log`:

```
Feb 22 13:36:10.555  INFO testinator: Installing rust toolchain 'nightly'
Feb 22 13:36:10.943  INFO testinator: Installing rust toolchain 'stable'
Feb 22 13:36:11.191  INFO testinator: Installing rust toolchain '1.29.0'
Feb 22 13:36:11.383  INFO testinator: Preparing environment for rust stable tests (8 configurations)
Feb 22 13:36:11.383  INFO testinator: Preparing environment for rust 1.29.0 tests (8 configurations)
Feb 22 13:36:11.384  INFO testinator: Preparing environment for rust nightly tests (9 configurations)
Feb 22 13:37:25.931  INFO testinator: Running rust nightly tests in /tmp/rust-bitcoin-nightly.DBBudwObJKaQ/rust-bitcoin
Feb 22 13:37:25.934  INFO testinator: Running rust stable tests in /tmp/rust-bitcoin-stable.qCyzHlx68R9I/rust-bitcoin
Feb 22 13:37:25.937  INFO testinator: Running rust 1.29.0 tests in /tmp/rust-bitcoin-1.29.0.AlaHDIqQLlAf/rust-bitcoin
Feb 22 13:37:25.937 DEBUG testinator: Generating lock file with rust=1.29.0
Feb 22 13:37:27.893 DEBUG testinator: Pinning cc to 1.0.41
Feb 22 13:37:28.062 DEBUG testinator: Pinning serde to 1.0.98
Feb 22 13:37:28.278 DEBUG testinator: Pinning serde_derive to 1.0.98
Feb 22 13:37:28.628 DEBUG testinator: Pinning byteorder to 1.3.4
Feb 22 13:38:23.294  INFO testinator: Test rust=nightly, features=[secp-recovery] succeeded!
Feb 22 13:38:25.660  INFO testinator: Test rust=stable, features=[secp-recovery] succeeded!
Feb 22 13:39:01.637  INFO testinator: Test rust=1.29.0, features=[secp-recovery] succeeded!
Feb 22 13:39:05.752  INFO testinator: Test rust=nightly, features=[use-serde] succeeded!
Feb 22 13:39:28.007  INFO testinator: Test rust=stable, features=[use-serde] succeeded!
Feb 22 13:39:34.832  INFO testinator: Test rust=nightly, features=[base64] succeeded!
Feb 22 13:39:57.354  INFO testinator: Test rust=stable, features=[base64] succeeded!
Feb 22 13:39:58.577  INFO testinator: Test rust=nightly, features=[secp-lowmemory] succeeded!
Feb 22 13:40:21.213  INFO testinator: Test rust=nightly, features=[rand] succeeded!
Feb 22 13:40:28.537  INFO testinator: Test rust=stable, features=[secp-lowmemory] succeeded!
Feb 22 13:40:47.660  INFO testinator: Test rust=nightly, features=[unstable] succeeded!
Feb 22 13:41:07.994  INFO testinator: Test rust=stable, features=[rand] succeeded!
Feb 22 13:41:08.000  INFO testinator: Test rust=1.29.0, features=[use-serde] succeeded!
Feb 22 13:41:35.112  INFO testinator: Test rust=nightly, features=[bitcoinconsensus] succeeded!
Feb 22 13:41:44.723  INFO testinator: Test rust=stable, features=[bitcoinconsensus] succeeded!
Feb 22 13:42:09.568  INFO testinator: Test rust=nightly, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,unstable,bitcoinconsensus] succeeded!                                                                                                  
Feb 22 13:42:11.081  INFO testinator: Test rust=1.29.0, features=[base64] succeeded!
Feb 22 13:42:22.465  INFO testinator: Test rust=stable, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Feb 22 13:42:34.005  INFO testinator: Test rust=nightly, features=[] succeeded!
Feb 22 13:42:50.333  INFO testinator: Test rust=stable, features=[] succeeded!
Feb 22 13:43:09.905  INFO testinator: Test rust=1.29.0, features=[secp-lowmemory] succeeded!
Feb 22 13:43:50.721  INFO testinator: Test rust=1.29.0, features=[rand] succeeded!
Feb 22 13:44:38.477  INFO testinator: Test rust=1.29.0, features=[bitcoinconsensus] succeeded!
Feb 22 13:45:26.189  INFO testinator: Test rust=1.29.0, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Feb 22 13:46:12.448  INFO testinator: Test rust=1.29.0, features=[] succeeded!
Feb 22 13:46:14.303  INFO testinator: Fuzzing deserialize_script
Feb 22 13:47:15.099  INFO testinator: Successfully fuzzed deserialize_script
Feb 22 13:47:15.100  INFO testinator: Fuzzing uint128_fuzz
Feb 22 13:48:16.205  INFO testinator: Successfully fuzzed uint128_fuzz
Feb 22 13:48:16.205  INFO testinator: Fuzzing deserialize_amount
Feb 22 13:49:17.119  INFO testinator: Successfully fuzzed deserialize_amount
Feb 22 13:49:17.119  INFO testinator: Fuzzing deserialize_transaction
Feb 22 13:50:18.245  INFO testinator: Successfully fuzzed deserialize_transaction
Feb 22 13:50:18.245  INFO testinator: Fuzzing deser_net_msg
Feb 22 13:51:19.223  INFO testinator: Successfully fuzzed deser_net_msg
Feb 22 13:51:19.223  INFO testinator: Fuzzing deserialize_address
Feb 22 13:52:20.136  INFO testinator: Successfully fuzzed deserialize_address
Feb 22 13:52:20.136  INFO testinator: Fuzzing deserialize_block
Feb 22 13:53:21.067  INFO testinator: Successfully fuzzed deserialize_block
Feb 22 13:53:21.067  INFO testinator: Fuzzing outpoint_string
Feb 22 13:54:22.170  INFO testinator: Successfully fuzzed outpoint_string
Feb 22 13:54:22.170  INFO testinator: Fuzzing deserialize_psbt
Feb 22 13:55:23.141  INFO testinator: Successfully fuzzed deserialize_psbt
```
</details>

**Review by apoelstra — APPROVED** [review]
ack 7b7b73f6d2653d42be5bd63aa01a07680a7539af

Looks good now :)

**Comment by stevenroose** [comment]
> As long as you take ownership of reverting the commit I'm ok with it, I'd just be unhappy with it staying that way and we don't exactly have a track record of being super organized as an org smile

@sgeisler we should use the issues feature more and have a release board or so at least. But well I'm not really the one to speak, I have been really bad keeping up with review lately.

**Review by stevenroose — APPROVED** [review]

**Pull request merged**
